src/server.rs

```diff
@@ -271,6 +310,63 @@ impl Server {
         }
     }
 
+    // Server browser related query packet processing for shared socket mode.  These are used
+    // when you pass STEAMGAMESERVER_QUERY_PORT_SHARED as the query port to SteamGameServer_Init.
+    // IP address and port are in host order, i.e 127.0.0.1 == 0x7f000001
+
+    // These are used when you've elected to multiplex the game server's UDP socket
+    // rather than having the master server updater use its own sockets.
+    //
+    // Source games use this to simplify the job of the server admins, so they
+    // don't have to open up more ports on their firewalls.
+
+    // Call this when a packet that starts with 0xFFFFFFFF comes in. That means
+    // it's for us.
+    pub fn handle_incoming_packet(
+        &self,
+        data: &mut [u8],
+        size: usize,
+        addr: u32,
+        port: u16,
+    ) -> bool {
+        unsafe {
+            let result = sys::SteamAPI_ISteamGameServer_HandleIncomingPacket(
+                self.server,
+                data.as_ptr() as _,
+                size as _,
+                addr,
+                port,
+            );
+            return result;
+        }
+    }
+
+    // AFTER calling HandleIncomingPacket for any packets that came in that frame, call this.
+    // This gets a packet that the master server updater needs to send out on UDP.
+    // It returns the length of the packet it wants to send, or 0 if there are no more packets to send.
+    // Call this each frame until it returns 0.
+    pub fn get_next_outgoing_packet(
```

Comment:
Use SocketAddrV4 here as well. Steamworks docs says the buffer should always be 16KB, so maybe it's more ergonomic to just return a Vec<u8> from here instead


```suggestion
    pub fn get_next_outgoing_packet(&self) -> Option<(SocketAddrV4, Vec<u8>)>
```

Reply:
Allocating on each packet read seems like a bad idea given how regularly this needs to be called, especially if the it may allocate upwards of 16KB each time.

Reply:
Hm, true. I don't think it would be a bad perf impact, but I guess it's good manner to not allocate.

Re-reading the docs for the SDK function it does mention that the size of the buffer "_should be 16 * 1024._". Maybe we add an assertion here for the size? 🤔 Also, 
> NOTE: This MUST be called repeatedly each frame until it returns 0 when in GameSocketShare mode.
I think this condition should be represented in the wrapper API. Callback is an option, what do you think?

```rs
/// Call this function after calling handle_incoming_packet. The callback
/// function `cb` will be called for each outgoing packet that needs to be sent
/// to an address. The buffer must be at least 16384 bytes in size.
pub fn get_next_outgoing_packet(&self, buffer: &mut [u8], cb: impl Fn(SocketAddrV4, &[u8])) {
    assert!(buffer.len() >= 16 * 1024, "Buffer size must be at least 16 KiB");

    loop {
        let mut addr = 0u32;
        let mut port = 0u16;

        let len = unsafe {
            sys::SteamAPI_ISteamGameServer_GetNextOutgoingPacket(
                self.server,
                buffer.as_mut_ptr() as *mut _,
                buffer.len() as _,
                &mut addr,
                &mut port,
            )
        };

        if len == 0 {
            break; // No more packets to process
        }

        let addr = SocketAddrV4::new(Ipv4Addr::from_bits(addr), port);
        cb(addr, &buffer[..len as usize]);
    }
}
```

Reply:
When using these changes, I provided a pre-allocated buffer that is reused each frame. Definitely doesn't feel right to alloc within the `get_next_outgoing_packet`. I should probably add an example of this use case to the example code -- It just ends up a little convoluted because `handle_incoming_packet`/`get_next_outgoing_packet` are meant for the scenario where your game server provides its own socket for connectivity instead of using the steam network.

Reply:
The packet may be up to maximum 16KiB, so the buffer should be _minimum_ 16KiB to hold that packet. The SDK doesn't specify what happens if the buffer cannot fit the packet it tries to send (say when the user provided a 1-byte buffer), and it doesn't have a way to tell us that the buffer provided was too small.

I dont know how badly the SDK will misbehave if you _don't_ handle all out going packets within the frame like it tells you to, so I'd like to be "strict" about it to the user. At bare minimum the docs should say that this must be polled to completion

Reply:
Implemented suggested solution, I don't have a particular issue with polling internally, just wasn't sure how batteries-included we wanted to be.